src/ssl.c

```diff
@@ -3448,6 +3448,207 @@ void wolfSSL_CertManagerFree(WOLFSSL_CERT_MANAGER* cm)
 
 }
 
+#if defined(OPENSSL_EXTRA) && !defined(NO_FILESYSTEM)
+#if defined(WOLFSSL_SIGNER_DER_CERT)
+/******************************************************************************
+* wolfSSL_CertManagerGetCerts - retrieve stack of X509 certificates in a
+* certificate manager (CM).
+*
+* RETURNS:
+* returns stack of X509 certs on success, otherwise returns a NULL.
+*/
+WOLFSSL_STACK* wolfSSL_CertManagerGetCerts(WOLFSSL_CERT_MANAGER* cm)
+{
+    WOLFSSL_STACK* sk = NULL;
+    Signer* signers = NULL;
+    word32  row = 0;
+    DecodedCert* dCert = NULL;
+    WOLFSSL_X509* x509 = NULL;
+    int found = 0;
+
+    if (cm == NULL)
+        return NULL;
+
+    sk = wolfSSL_sk_X509_new();
+
+    if (sk == NULL) {
+        return NULL;
+    }
+
+    if (wc_LockMutex(&cm->caLock) != 0) {
+        goto error_init;
+    }
+
+    for (row = 0; row < CA_TABLE_SIZE; row++) {
+        signers = cm->caTable[row];
+        while (signers && signers->derCert && signers->derCert->buffer) {
+
+            dCert = (DecodedCert*)XMALLOC(sizeof(DecodedCert), cm->heap,
```

Comment:
This should only need to be allocated and free'd once. Otherwise it creates possible heap fragmentation... Also only do the allocation for `DecodedCert` in the `WOLFSSL_SMALL_STACK` case. Here is an example:

```c
#ifdef WOLFSSL_SMALL_STACK
    DecodedCert* dCert = NULL;
#else
    DecodedCert  dCert[1];
#endif

#ifdef WOLFSSL_SMALL_STACK
    dCert = (DecodedCert*)XMALLOC(sizeof(DecodedCert), NULL,
                                                   DYNAMIC_TYPE_DCERT);
    if (dCert != NULL) {
        /* error */
    }
#endif
    XMEMSET(dCert, 0, sizeof(DecodedCert));
```

Reply:
@dgarske, heap fragmentation? I don't think I can use static 'dCert[1]' since ParseCert() may allocate additional memory for dCert and having malloc()/free() cleans it nicely.

Reply:
On embedded systems we try and avoid malloc/free over and over as it can cause heap fragmentation. It's not something we think about with desktop development, but it's a frequent issue on embedded systems. This pattern I showed above is used in many places in ssl.c it would be good to keep this format. Move the alloc/free out of the loop so its only done once. The `DecodedCert  dCert[1];` is just a trick so the pointer reference is the same with or without small stack.

Reply:
@dgarske, In this case, the same block size is allocated and released so I don't see how it would cause heap fragmentations.